_LowLevelCode/cpp/cpp_communicator/serialiser/c_serialise.cpp

```diff
@@ -0,0 +1,292 @@
+/*=========================================================
+ * c_serialise.cpp
+ * Serialise MATLAB object into a uint8 data stream
+ *
+ * See also:
+ * hlp_serialize
+ * hlp_deserialize
+ *
+ * This is a MEX-file for MATLAB.
+ *=======================================================*/
+
+#include <iostream>
+#include <cstring>
+#include <cmath>
+#include <vector>
+#include "cpp_serialise.hpp"
+
+template<typename T>
+inline void ser(uint8_t* data, size_t& memPtr, const std::vector<T> data_in, const size_t amount) {
+  memcpy(&data[memPtr], data_in.data(), amount);
+  memPtr += amount;
+}
+
+inline void ser(uint8_t* data, size_t& memPtr, const void* const data_in, const size_t amount) {
+  // Write bytes and move memory index
+  memcpy(&data[memPtr], data_in, amount);
+  memPtr += amount;
+}
+
+inline void write_data(uint8_t* data, size_t& memPtr, const mxArray* const input, const size_t elemSize, const size_t nElem) {
+  if (mxIsComplex(input)) {
+    // Size of a complex component is half that of the whole complex
+    size_t compSize = elemSize/2;
+
+#if MX_HAS_INTERLEAVED_COMPLEX
+    void* toWrite = mxGetComplexDouble(input);
+    // Offset imaginary to end
+    size_t imPtr = memPtr + (1+nElem)*compSize;
+
+    for (int i = 0; i < nElem; i+=elemSize, memPtr += compSize, imPtr += compSize) {
+      memcpy(&data[memPtr], &toWrite[i], compSize);
+      memcpy(&data[imPtr], &toWrite[i+compSize], compSize);
+    }
+    memPtr = imPtr; // Finally update memPtr
+
+#else
+    void* toWrite = mxGetPr(input);
+    ser(data, memPtr, toWrite, compSize*nElem);
+    toWrite = mxGetPi(input);
+    ser(data, memPtr, toWrite, compSize*nElem);
+
+#endif
+
+  } else {
+    void* toWrite = mxGetPr(input);
+    ser(data, memPtr, toWrite, elemSize*nElem);
+  }
+}
+
+inline void write_header(uint8_t* data, size_t& memPtr, tag_type& tag,
+                         const size_t nElem, const mwSize* dims, const size_t nDims) {
+
+  if (nElem == 0) { // Null
+    tag.dim = 1;
+    ser(data, memPtr, &tag, 1);
+    ser(data, memPtr, &nElem, 4);
+  }
+  else if (nElem == 1) { // Scalar
+    tag.dim = 0;
+    ser(data, memPtr, &tag, 1);
+  }
+  else if (nDims == 2 && dims[0] == 1) { // List
+    tag.dim = 1;
+    ser(data, memPtr, &tag, 1);
+    ser(data, memPtr, &nElem, types_size[UINT32]);
+  }
+  else { // General array
+    tag.dim = nDims;
+
+    std::vector<uint32_t> cast_dims(nDims);
+    for (int i = 0; i < nDims; i++) cast_dims[i] = (uint32_t) dims[i];
+
+    ser(data, memPtr, &tag, 1);
+    ser(data, memPtr, cast_dims, nDims*types_size[UINT32]);
+  }
+
+}
+
+
+void serialise(uint8_t* data, size_t& memPtr, const mxArray* input){
+
+  tag_type tag = tag_data(input);
+
+  const size_t nElem = mxGetNumberOfElements(input);
+  const mwSize* dims = mxGetDimensions(input);
+  const size_t nDims = mxGetNumberOfDimensions(input);
+
+  for (int i=0; i < nDims; i++) {
+    if (dims[i] > DIM_MAX) {
```

Comment:
what datatype is the dims array? Can it contain numbers `>max(uint32)`?

Reply:
It might be `size_t`, though the docs say it can change, so in principle, yes. Hence the checks.